process-engine/process_engine_runtime
# ✨ Socket io notifications

**Changes:**

Implement socket.io notifications for the ManagementAPI and ConsumerAPI.

With this first implementation, four global channels are supported:
- `user_task_reached` - A UserTask has been reached and is now waiting to be handled (i.e. in a suspended state).
- `user_task_finished` - A UserTask has been finished.
- `process_terminated` - A process instance was prematurely aborted by a `TerminateEndEvent`.
- `process_ended` - A process instance has finished execution.

Each message published through either of these channels contains the following information:
- `correlationId` - The ID of the correlation the ProcessIntance belongs to.
- `processInstanceId` - The ID of the ProcessInstance that is/was executed.
- `processModelId` - The ID of the ProcessModel that describes the ProcessInstance.
- `flowNodeId` - The ID of the FlowNode that triggered the event (i.e. EndEventId or UserTaskId, based on the currently supported channels).
- `flowNodeInstanceId` - The InstanceId of the FlowNode that triggered the Event.
- `currentToken` - The ProcessToken belonging to the FlowNode.

**Issues:**

Closes #71 

PR: #87

## How can others test the changes?

You can use the `ConsumerApiClient` or the `ManagementApiClient` to subscribe to any one of the mentioned channels.
Each has its own subscription method for this purpose.

See:
[Management API](https://github.com/process-engine/management_api_contracts/blob/develop/src/imanagement_api.ts#L325)
[Consumer API](https://github.com/process-engine/consumer_api_contracts/blob/develop/typescript/src/iconsumer_api.ts#L251)

Simply pass the callback function you wish to be called when receiving a notification to the respective method .

## PR-Checklist

Please check the boxes in this list after submitting your PR:


## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>
